linkerd/core/src/main/scala/io/buoyant/linkerd/Linker.scala

```diff
@@ -1,175 +1,94 @@
-   * protocol support.
-   *
-   * A linker configuration consists of the following:
-   *  - arbitrary default [[Router.Params router params]]
-   *  - a required `routers` field with a list of [[Router]] specs
-   *
-   * Example configuration:
-   *
-   * <pre>
-   *    admin:
-   *      port: 9991
-   *
-   *     baseDtab: |
-   *       /bar => /bah;
-   *       /http/1.1 => /b;
-   *
-   *     routers:
-   *      - protocol: http
-   *        servers:
-   *        - port: 4140
-   *        httpUriInDst: true
-   *
-   *      - protocol: http
-   *        label: http.ext
-   *        dstPrefix: /ext/http
-   *        servers:
-   *        - port: 8080
-   *          ip: any
-   *
-   *      - protocol: thrift
-   *        client:
-   *          thriftFramed: true
-   *        servers:
-   *        - port: 9090
-   *
-   * </pre>
-   *
-   * Note that protocol-specific features are delegated to per-
-   * protocol [[ProtocolInitializer]] plugins.
-   *
-   * Ordering of fields is irrelevant.
-   */
-  def read(p: JsonParser): Linker
+  def configured[T: Stack.Param](t: T): Linker
 }
 
 object Linker {
 
-  def mk(protos: ProtocolInitializers, namers: NamerInitializers, tls: TlsClientInitializers): Linker =
-    Impl(protos, namers, tls, Stack.Params.empty, Nil, Admin())
+  def parse(config: String, configInitializers: Seq[ConfigInitializer]): LinkerConfig = {
+    val mapper = Parser.objectMapper(config)
+    // Subtypes must not conflict
+    configInitializers.groupBy(_.configId).collect {
+      case (id, cis) if cis.size > 1 =>
+        throw ConflictingSubtypes(cis(0).namedType, cis(1).namedType)
+    }
+    for (ci <- configInitializers) ci.registerSubtypes(mapper)
+    // TODO: Store the LinkerConfig so that it can be serialized out later
+    mapper.readValue[LinkerConfig](config)
+  }
+
+  def load(config: String, configInitializers: Seq[ConfigInitializer]): Linker = {
+    parse(config, configInitializers).mk
+  }
+
+  def parse(config: String): LinkerConfig = {
+    val protocols = LoadService[ProtocolInitializer]
+    val namers = LoadService[NamerInitializer]
+    val clientTls = LoadService[TlsClientInitializer]
+    parse(config, protocols ++ namers ++ clientTls)
+  }
+
+  def load(config: String): Linker = {
+    parse(config).mk
+  }
+
+  case class LinkerConfig(namers: Option[Seq[NamerConfig]], routers: Seq[RouterConfig], admin: Option[Admin]) {
+    def mk: Linker = {
+      val namersParam = Stack.Params.empty.maybeWith(namers.map(nameInterpreter).map(DstBindingFactory.Namer(_)))
 
-  def load(): Linker = {
-    val protos = ProtocolInitializers.load()
-    val namers = NamerInitializers.load()
-    val tls = TlsClientInitializers.load()
-    mk(protos, namers, tls)
+      // At least one router must be specified
+      if (routers.isEmpty) {
+        throw NoRoutersSpecified
+      }
+
+      // Router labels must not conflict
+      routers.groupBy(_.label).collect {
```

Comment:
nit: since we're only doing this for the side-effects, I'd prefer a `foreach` to `collect` here.